src/hb-subset-plan.cc

```diff
@@ -506,16 +507,65 @@ _create_old_gid_to_new_gid_map (const hb_face_t *face,
 
 static void
 _nameid_closure (hb_face_t *face,
-		 hb_set_t  *nameids)
+		 hb_set_t  *nameids,
+		 bool all_axes_pinned,
+		 hb_hashmap_t<hb_tag_t, float> *user_axes_location)
 {
 #ifndef HB_NO_STYLE
-  face->table.STAT->collect_name_ids (nameids);
+  face->table.STAT->collect_name_ids (user_axes_location, nameids);
 #endif
 #ifndef HB_NO_VAR
-  face->table.fvar->collect_name_ids (nameids);
+  if (!all_axes_pinned)
+    face->table.fvar->collect_name_ids (user_axes_location, nameids);
 #endif
 }
 
+
+static void
+_normalize_axes_location (hb_face_t *face,
+			  hb_hashmap_t<hb_tag_t, int> *normalized_axes_location,
+			  hb_hashmap_t<hb_tag_t, float> *user_axes_location,
+			  bool &all_axes_pinned,
+			  bool &pinned_at_default)
+{
+  if (user_axes_location->is_empty ())
+    return;
+
+  hb_blob_ptr_t<OT::fvar> fvar = hb_sanitize_context_t ().reference_table<OT::fvar> (face);
+  unsigned axisCount = fvar->get_axis_count ();
+  hb_array_t<const OT::AxisRecord> axes = fvar->get_axes ();
+  fvar.destroy ();
+
+  hb_blob_ptr_t<OT::avar> avar = hb_sanitize_context_t ().reference_table<OT::avar> (face);
+  bool has_avar = avar->has_data ();
+  const OT::SegmentMaps *seg_maps = nullptr;
+  if (has_avar)
+    seg_maps = avar->get_segment_maps ();
+
+  avar.destroy ();
+  bool axis_not_pinned = false;
+  for (unsigned i = 0; i < axisCount; i++)
+  {
+    hb_tag_t axis_tag = axes[i].get_axis_tag ();
+    if (!user_axes_location->has (axis_tag))
+    {
+      axis_not_pinned = true;
+    }
+    else
+    {
+      int normalized_v = axes[i].normalize_axis_value (user_axes_location->get (axis_tag));
+      if (has_avar)
+      {
+        normalized_v = seg_maps->map (normalized_v);
+      }
+      normalized_axes_location->set (axis_tag, normalized_v);
+      if (normalized_v != 0) pinned_at_default = false;
+    }
+    if (has_avar)
+      seg_maps = &StructAfter<OT::SegmentMaps> (*seg_maps);
```

Comment:
Although avar is suposed to have the same number of axes as fvar, in a malformed font it may not. You'll need to check you don't exceed the axisCount specified on avar.